synapse/storage/schema/main/delta/78/01_validate_and_update_profiles.py

```diff
@@ -0,0 +1,85 @@
+# Copyright 2023 The Matrix.org Foundation C.I.C
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from synapse.config.homeserver import HomeServerConfig
+from synapse.storage.database import LoggingTransaction
+from synapse.storage.engines import BaseDatabaseEngine, PostgresEngine
+
+
+def run_upgrade(
+    cur: LoggingTransaction,
+    database_engine: BaseDatabaseEngine,
+    config: HomeServerConfig,
+) -> None:
+    hostname = config.server.server_name
+
+    if isinstance(database_engine, PostgresEngine):
+        # check if the constraint can be validated
+        check_sql = """
+                    SELECT user_id from profiles WHERE full_user_id IS NULL
+                    """
+        cur.execute(check_sql)
+        res = cur.fetchall()
+
+        if res:
+            # there are rows the background job missed, finish them here before we validate the constraint
+            process_rows_sql = """
+                               UPDATE profiles
+                               SET full_user_id = '@' || user_id || ?
```

Comment:
Perhaps people are familiar with the `||` string concatenation operator but I had to look it up.

I noticed there was a `CONCAT(...)` function we could use which reads more straight-forward to me. (up to you)

Reply:
I used this syntax as `||` is valid in both SQLite and Postgres but SQLite doesn't recognize `CONCAT(...)` and it felt easier to standardize it across both.